bucket-notification.go

```diff
@@ -0,0 +1,144 @@
+	ObjectCreatedPost                          = "s3:ObjectCreated:Post"
+	ObjectCreatedCopy                          = "s3:ObjectCreated:Copy"
+	ObjectCreatedCompleteMultipartUpload       = "sh:ObjectCreated:CompleteMultipartUpload"
+	ObjectRemovedAll                           = "s3:ObjectRemoved:*"
+	ObjectRemovedDelete                        = "s3:ObjectRemoved:Delete"
+	ObjectRemovedDeleteMarkerCreated           = "s3:ObjectRemoved:DeleteMarkerCreated"
+	ObjectReducedRedundancyLostObject          = "s3:ReducedRedundancyLostObject"
+)
+
+type FilterRule struct {
+	XMLName xml.Name `xml:"FilterRule,omitempty"`
+	Name    string   `xml:"Name"`
+	Value   string   `xml:"Value"`
+}
+
+type S3Key struct {
+	FilterRules []FilterRule
+}
+
+type Filter struct {
+	S3Key S3Key `xml:"S3Key,omitempty"`
+}
+
+// Arn - holds ARN information that will be sent to the web service
+type Arn struct {
+	Partition string
+	Service   string
+	Region    string
+	AccountID string
+	Resource  string
+}
+
+func NewArn(partition, service, region, accountID, resource string) Arn {
+	return Arn{Partition: partition,
+		Service:   service,
+		Region:    region,
+		AccountID: accountID,
+		Resource:  resource}
+}
+
+func (arn Arn) String() string {
+	return "arn:" + arn.Partition + ":" + arn.Service + ":" + arn.Region + ":" + arn.AccountID + ":" + arn.Resource
+}
+
+// NotificationConfig - represents one single notification configuration
+// such as topic, queue or lambda configuration.
+type NotificationConfig struct {
+	Id     string  `xml:"Id,omitempty"`
+	Arn    Arn     `xml:"-"`
+	Events []Event `xml:"Event"`
+	Filter *Filter `xml:"Filter,omitempty"`
+}
+
+func NewNotificationConfig(arn Arn) NotificationConfig {
+	return NotificationConfig{Arn: arn}
+}
+
+func (t *NotificationConfig) AddEvents(events ...Event) {
+	t.Events = append(t.Events, events...)
+}
+
+func (t *NotificationConfig) AddFilterSuffix(suffix string) {
+	if t.Filter == nil {
+		t.Filter = &Filter{}
+	}
+	t.Filter.S3Key.FilterRules = append(t.Filter.S3Key.FilterRules, FilterRule{Name: "suffix", Value: suffix})
+}
+
+func (t NotificationConfig) AddFilterPrefix(prefix string) {
+	if t.Filter == nil {
+		t.Filter = &Filter{}
+	}
+	t.Filter.S3Key.FilterRules = append(t.Filter.S3Key.FilterRules, FilterRule{Name: "prefix", Value: prefix})
+}
+
+// Topic notification config
+type TopicConfig struct {
+	NotificationConfig
+	XMLName xml.Name `xml:"TopicConfiguration"`
+	Topic   string   `xml:"Topic"`
+}
+
+type QueueConfig struct {
+	NotificationConfig
+	XMLName xml.Name `xml:"QueueConfiguration"`
+	Queue   string   `xml:"Queue"`
+}
+
+type LambdaConfig struct {
+	NotificationConfig
+	XMLName xml.Name `xml:"CloudFunctionConfiguration"`
+	Lambda  string   `xml:"CloudFunction"`
+}
+
+// BucketNotification - the struct that represents the whole XML to be sent to the web service
+type BucketNotification struct {
+	XMLName       xml.Name `xml:"NotificationConfiguration"`
+	LambdaConfigs []LambdaConfig
```

Comment:
This should be - https://github.com/aws/aws-sdk-go/blob/master/service/s3/api.go#L6992